### PR TITLE
BFD-1010: Create new BFD databases table structure, migrate data from current tables

### PR DESCRIPTION
**Pull request by mackec1**
**JIRA Ticket:**
[BFD-1010](https://jira.cms.gov/browse/BFD-1010)

**User Story or Bug Summary:**
This ticket will create a new BFD table structure that uses table column names more closely aligned with CCW column names. Current table data will be migrated from, and verified vs., current production data.

### What Does This PR Do?
This migration has two primary goals:

1. convert the beneficiaryId, claimId and claimGroupId columns to `bigint`, rather than their current `varchar(11)` types. Per the PostgreSQL documentation, a `varchar(11)` column requires 16 bytes per record on disk, and a `bigint` column only requires 8 bytes. Accordingly, this change roughly halves the size of our indexes using those columns.
2.  _finally_ moving to a more idiomatic (and user-friendly) naming scheme for these tables based (almost entirely) on CCW field names.

In addition to those primary goals, this schema migration is also being used as a good opportunity to address all of these other less major schema changes:

- Use better defined Postgres data types (i.e., NUMERIC(3,0) to SMALLINT)
- Minor changes to column ordering to facilitate potential indexing
- Group together like data elements.

In all table definitions, there is a comment identifier to the right of each table column; this represents the previous schema name for that data column.

### What Should Reviewers Watch For?
There are three main components (sections) that reviewers should focus on:

1. V37__Create_Schema_Idiomatic_Names.sql which creates the new db table structure; check on appropriate use of correct data type(s), etc. Current db table column name is provided as a sql comment to the right of each column definition.
2. check all insert*.sql scripts for appropriate migration of current table data to the new formatted table name/structure.
3. check all verify*.sql scripts to ensure completeness and coverage.

There is also a checklist / instructions provided in the file: **MIGRATION_NOTES.txt**

Both the insert and verify scripts follow a pattern that should be easy to follow; the reviewer should be able to see the pattern, especially for the verification of data steps.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

* N/A

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql [hunk 1]
```diff
@@ -0,0 +1,1787 @@
...
+    bene_crnt_hic_num                        character varying(64) not null,           -- hicn
+    mbi_num                                  character varying(11),                    -- medicareBeneficiaryId
+    hicn_unhashed                            character varying(11),                    -- hicnUnhashed
+    constraint beneficiaries_history_invalid_beneficiaries_pkey
+        primary key (bene_history_id)
+);
+
+create table if not exists beneficiary_monthly (
+    year_month                               date not null,                            -- yearMonth
+    parent_beneficiary                       bigint not null,                          -- parentBeneficiary
```
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql:299]: any thoughts of changing this to bene_id?

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql [hunk 2]
```diff
@@ -0,0 +1,1787 @@
...
+    drvd_line_6_adr                          character varying(40),                    -- derivedMailingAddress6
+    city_name                                character varying(100),                   -- derivedCityName
+    state_cd                                 character varying(2),                     -- derivedStateCode
+    state_cnty_zip_cd                        character varying(9),                     -- derivedZipCode
+    bene_link_key                            bigint,                                   -- beneLinkKey
+    constraint beneficiaries_pkey
+        primary key (bene_id)
+);
+
+create index if not exists beneficiaries_hicn_idx
```
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql:252]: for consistency should this be called beneficiaries_bene_crnt_hic_num_idx?

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql (hunk 1 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql:299]: thoughts by me...for sure! BUT, for wahtever reason, we have db columns using that 'parent_' naming convention (i.e., parent_claim). My pref would be to scrap that use of 'parent_' and use table-qualified names as needed; for example...
`
select d.* from carrier_claim_lines d
inner join carrier_claims c on c.clm_id = d.clm_id
inner join beneficiaries b where b.bene_id = c.bene_id
`

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql (hunk 2 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql:252]: did some renaming to try to adhere to your suggestion; where we run into issue(s) is that the index and/or constraint names are limited to 64 characters. Thus, we will always have a case where the name can't quite 'fit' into 64 chars; for example:
`
create index if not exists beneficiary_monthly_yearmonth_partdcontract_parentbeneficiary_idx
	on beneficiary_monthly (year_month, partd_contract_number_id asc, parent_beneficiary asc);
`

will get truncated to 64 chars. Now I am open to any/all suggestions as to how these things could be named, but at the end of the day, it's just a property name that needs to be unique within the schema.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql [hunk 3]
```diff
@@ -0,0 +1,1787 @@
...
+
+create index if not exists beneficiaries_history_hicn_idx
+	on beneficiaries_history (bene_crnt_hic_num);
+
+create index if not exists beneficiaries_history_mbi_hash_idx
+	on beneficiaries_history (mbi_hash);
+
+create table if not exists beneficiaries_history_invalid_beneficiaries (
+    bene_history_id                          bigint not null,                          -- beneficiaryHistoryId
+    bene_id                                  bigint,                                   -- beneficiaryId
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__Create_Schema_Idiomatic_Names.sql:287]: good questions, one I don't know the answer to. The current (prod) schema allows for the "beneficiaryId" to be null, so I complied with that. Perhaps the naming of the table (i.e., ..._invalid_beneficiaries) implies that a null "beneficiaryId" is fair game.

**Comment by christrotteradhoc** [comment]
can we close this PR now since we are breaking it up?

**Pull request closed without merging**